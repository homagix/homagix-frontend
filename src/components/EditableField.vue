<script setup lang="ts">
import { computed } from 'vue';


const props = defineProps<{
  type?: "string" | "number"
  value: string | number
  editable: boolean
}>()

const emit = defineEmits<{
  (event: "input", value: string): void
}>()

function input(event: Event) {
  emit("input", (event.target as HTMLInputElement).innerText)
}

const classes = computed(() => props.type && ({ [props.type]: true }))
</script>

<template>
  <span
    :class="classes"
    :contenteditable="editable"
    @input="input"
  >
    {{ value }}
  </span>
</template>

<style lang="scss" scoped>
span {
  padding: 2px 4px;

  &.editable {
    border-color: auto;
    appearance: inherit;
    width: auto;
  }

  &.number {
    text-align: right;
    background-color: transparent;
    border-radius: 0;
    display: inline;
    font-size: inherit;
    margin: 0;
    min-width: unset;
    height: auto;
  }
}
</style>
