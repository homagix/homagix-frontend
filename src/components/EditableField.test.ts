import { mount } from "@vue/test-utils"
import { describe, expect, it } from "vitest"
import EditableField from "@/components/EditableField.vue"

describe("EditableField", () => {
  it("should render non-editable", () => {
    const wrapper = mount(EditableField, { props: { value: "abc", editable: false } })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render editable", () => {
    const wrapper = mount(EditableField, { props: { value: "abc", editable: true } })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render numeric fields", () => {
    const wrapper = mount(EditableField, { props: { type: "number", value: 123, editable: true } })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should emit an 'input' event when content was changed", async () => {
    const wrapper = mount(EditableField, { props: { value: "abc", editable: true } })
    const el = wrapper.find("span")
    el.element.innerText = "def"
    await el.trigger("input")
    expect(wrapper.emitted().input).toHaveLength(1)
    expect(wrapper.emitted().input[0]).toEqual(["def"])
  })
})
