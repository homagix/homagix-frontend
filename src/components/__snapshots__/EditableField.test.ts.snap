// Vitest Snapshot v1

exports[`EditableField > should render editable 1`] = `"<span class=\\"\\" contenteditable=\\"true\\" data-v-09beb9f6=\\"\\">abc</span>"`;

exports[`EditableField > should render non-editable 1`] = `"<span class=\\"\\" contenteditable=\\"false\\" data-v-09beb9f6=\\"\\">abc</span>"`;

exports[`EditableField > should render numeric fields 1`] = `"<span class=\\"number\\" contenteditable=\\"true\\" data-v-09beb9f6=\\"\\">123</span>"`;
